### PR TITLE
Add ability to set preferred name on onboarding + account page

### PR DESCRIPTION
**Pull request by samaratrilling**
This PR does the real work to implement the new preferred name feature on all our onboarding flows and the account settings page.
[ch6110]
Mocks: https://www.figma.com/file/GAwij1NO8C5UHNazo7sJyC/Account-Pages?node-id=33%3A2

Screen Shot 2021-06-28 at 2 21 34 PM

Screen Shot 2021-06-28 at 2 20 25 PM

Screen Shot 2021-06-28 at 2 18 53 PM

### CODE REVIEW

File: frontend/lib/common-steps/ask-nyc-address.tsx [hunk 1]
```diff
@@ -59,6 +59,7 @@ function getInitialState(s: AllSessionInfo): OnboardingStep1Input {
   return {
     firstName: "ignore",
     lastName: "ignore",
+    preferredFirstName: "ignore", // not sure if this should change
```
Comment (toolness) [inline frontend/lib/common-steps/ask-nyc-address.tsx:62]: Urg I think you can just set this to an empty string.  We really need to just make this form use its own separate mutation, rather than reusing onboarding step 1's mutation.  But since onboarding step 1 has preferred first name as an optional/empty field, this can probably just default to an empty string (it won't make the GraphQL endpoint spit out a validation error).

File: onboarding/forms.py [hunk 2]
```diff
@@ -82,6 +82,15 @@ class OnboardingStep1Form(AptNumberWithConfirmationForm, AddressAndBoroughFormMi
...
 
+    preferred_first_name = forms.CharField(
+        max_length=150,
+        required=False,
+        help_text=(
+            "The first name Justfix will call the user by. Optional. May be different from "
+            "their legal first name."
+        ),
+    )
+
```
Comment (toolness) [inline onboarding/forms.py:93]: Hmm, so you might actually need to make a duplicate version of this called `OnboardingStep1FormV2` or somesuch--it's possible that due to the way that these forms are translated to GraphQL, old versions of the front-end that don't specify any value for preferred first name (as opposed to specifying an empty string for it) will make the GraphQL endpoint explode.  😞

File: frontend/lib/common-steps/ask-name.tsx [hunk 3]
```diff
@@ -39,6 +40,26 @@ export const AskNameStep: React.FC<MiddleProgressStepProps> = (props) => {
...
+        {(ctx) => (
+          <>
+            <TextualFormField
+              {...ctx.fieldPropsFor("preferredFirstName")}
+              label={li18n._(t`Preferred first name`)}
+            />
+            <ProgressButtons isLoading={ctx.isLoading} back={props.prevStep} />
+          </>
+        )}
+      </SessionUpdatingFormSubmitter>
```
Comment (toolness) [inline frontend/lib/common-steps/ask-name.tsx:62]: I think we discussed this on a video call at some point, but because this is a second `<SessionUpdatingFormSubmitter>`, it effectively means that it needs to be submitted separately from the "legal first name" form... but these steps are constructed in such a way that they expect only _one_ submittable form per step, so once the user fills out just one of these forms, they will immediately be moved to the next step without having a chance to fill out the other form.

I recommend that instead of making a separate form, you either make a `NorentFullAndPreferredNameMutation` that supersedes `NorentFullNameMutation` (deprecating the latter), or you just put the preferred name on an actual separate step (i.e. page).

File: onboarding/schema.py [hunk 4]
```diff
@@ -92,10 +92,12 @@ def pick_model_fields(model, **kwargs):
 
 def complete_onboarding(request, info, password: Optional[str]) -> JustfixUser:
     with transaction.atomic():
+        print(info)
```
Comment (toolness) [inline onboarding/schema.py:95]: Just a reminder to remove this before merging!

File: onboarding/schema.py [hunk 5]
```diff
@@ -92,10 +92,12 @@ def pick_model_fields(model, **kwargs):
...
         user = JustfixUser.objects.create_user(
             username=JustfixUser.objects.generate_random_username(),
             first_name=info["first_name"],
             last_name=info["last_name"],
+            preferred_first_name=info["preferred_first_name"],
```
Comment (toolness) [inline onboarding/schema.py:100]: So above in line 24 of this file, you incremented `FIELD_SCHEMA_VERSION`, but a downside of that is that it will essentially "reset" the onboarding process for anyone who is currently in the middle of it, resulting in confusing behavior where all the information from their previous steps would suddenly disappear.  (Which is better than the server 500'ing, but it's still not ideal.)

I _think_ that an alternative to incrementing `FIELD_SCHEMA_VERSION` may be to simply assume that `preferred_first_name` might not actually be present in `info`.  That is, changing the above line to:

```python
preferred_first_name=info.get("preferred_first_name", ""),
```

might do the trick.  It would essentially make it so that anyone currently in the midst of onboarding would be akin to a user who chose not to fill in that field.

File: users/models.py [hunk 6]
```diff
@@ -148,6 +156,10 @@ def full_legal_name(self) -> str:
             return " ".join([self.first_name, self.last_name])
         return ""
 
+    @property
+    def preferred_name(self) -> str:
+        return self.preferred_first_name if self.preferred_first_name else self.first_name
```
Comment (toolness) [inline users/models.py:161]: This would be a good, super quick thing to add tests for in `users/tests/test_models.py`!  It could also be independent of the database, which would make the tests run particularly fast.

File: norent/scaffolding.py [hunk 7]
```diff
@@ -9,7 +9,7 @@
 
 
 # This should change whenever our scaffolding model's fields change.
-VERSION = "1"
+VERSION = "2"
```
Comment (toolness) [inline norent/scaffolding.py:12]: Have you tried _not_ changing this?  Because of how the norent scaffolding uses `pydantic.BaseModel` (unlike the onboarding scaffolding, which just stores raw dictionaries), I think it's possible that you might be able to get away with not changing this (and thereby not disrupting anyone who's currently onboarding) since `preferred_first_name` would just default to the empty string on the server side.  See my other comment about the onboarding `FIELD_SCHEMA_VERSION` for more details on why it'd be preferable to not have to increment this if possible.

**Review by toolness — COMMENTED** [review]
Hey, since you mentioned that you wanted to make sure this PR got merged soon, I went ahead and made some comments on it--I hope that's okay!  I probably mentioned some things that you already knew about but feel free to just consider them reminders...

File: data_driven_onboarding/data-driven-onboarding.sql [hunk 8]
```diff
@@ -122,7 +122,7 @@ Complaint_Category as(
     where bbl= %(bbl)s
     group by category
     order by NumberOfComplaints desc
-    limit 1
+    limit 1 -- maybe try using MAX here
```
[the remark this answers is not in the document]
Comment (samaratrilling) [reply data_driven_onboarding/data-driven-onboarding.sql:125]: Whoops, how'd that sneak in there...

File: frontend/lib/account-settings/about-you-settings.tsx [hunk 9]
```diff
@@ -5,13 +5,54 @@ import { AppContext } from "../app-context";
...
+    "preferredname"
+  );
+
+  return (
+    <>
+      {sec.heading}
+      <p>The name you'd like people to call you by.</p>
+      <EditableInfo
+        {...sec}
+        readonlyContent={`${session.preferredFirstName}`}
```
[the remark this answers is not in the document]
Comment (samaratrilling) [reply frontend/lib/account-settings/about-you-settings.tsx:27]: Interesting. When I try to shorten this to `{session.preferredFirstName}` it complains that readonlyContent can't have a null value at all.

`session.firstName` and `session.lastName` are both nullable, so I think we should keep `session.preferredFirstName`'s nullability, but I'm interested how to format this string to convert a `null` into an empty string here.
Ideally we'd like this to be an empty string

File: frontend/lib/account-settings/route-info.ts [hunk 10]
```diff
@@ -4,7 +4,8 @@ export function createAccountSettingsRouteInfo(prefix: string) {
   return {
     [ROUTE_PREFIX]: prefix,
     home: prefix,
-    name: `${prefix}/name`,
+    legalname: `${prefix}/legalname`,
+    preferredname: `${prefix}/preferredname`,
```
[the remark this answers is not in the document]
Comment (samaratrilling) [reply frontend/lib/account-settings/route-info.ts:8]: Yes, I remember that conversation - I just don't know if I got around to implementing it! Let's talk about this tomorrow - this will have implications for the whole `SessionUpdatingFormSubmitter` thing about whether there should be two or one.

**Comment by samaratrilling** [comment]
The goal here is to wait until we have a rollout plan that notifies users what this change is.
[ ] program team should be notified about how to talk with users about preferred name
[ ] write a plan for the metrics we want to track - we should make sure this doesn'tdecrease user throughput.

**Comment by samaratrilling** [comment]
The !importants on the CSS are definitely not right - would love a tip on why they're getting overridden by other CSS classes.

File: frontend/sass/styles.scss [hunk 11]
```diff
@@ -166,3 +166,10 @@ html.jf-is-fullscreen-admin-page {
     }
   }
 }
+
+.jf-account-settings-h3 {
+  font-weight: 300 !important;
+  font-size: 2rem !important;
```
Comment (toolness) [inline frontend/sass/styles.scss:172]: I think you said that you should get rid of these?

File: onboarding/forms.py [hunk 12]
```diff
@@ -16,12 +16,12 @@
...
 
 # Whenever we change the fields in any of the onboarding
-# forms, we should change this number to ensure that we
+# forms, we should increment this number to ensure that we
 # never use an old session's onboarding data with the
 # new validation logic. The downside is that the old
 # session's onboarding data will disappear, but hopefully
 # we won't have to do this often.
-FIELD_SCHEMA_VERSION = 4
+FIELD_SCHEMA_VERSION = 5
```
Comment (toolness) [inline onboarding/forms.py:24]: Hmm, it's possible that we don't need to increment this now that we are manually migrating the old schema to the new one (see the `migrate_dict` method in this PR)...  Which I thought was the reason we added that auto-migration functionality (otherwise incrementing this number means that people who are in the middle of onboarding will suddenly have all their onboarding data cleared, which would be confusing).

File: users/models.py [hunk 13]
```diff
@@ -149,6 +157,10 @@ def full_legal_name(self) -> str:
...
```
Comment (toolness) [inline users/models.py:162]: Err wait I remember being a bit confused by this--the name `preferred_name` makes it sound as though this might be the preferred _full_ name of the user rather than the preferred _first_ name of the user (the implementation suggests it's the latter), what do you think?

File: frontend/lib/account-settings/route-info.ts [hunk 14]
```diff
@@ -4,7 +4,8 @@ export function createAccountSettingsRouteInfo(prefix: string) {
   return {
     [ROUTE_PREFIX]: prefix,
     home: prefix,
-    name: `${prefix}/name`,
+    name: `${prefix}/name`, // This is name rather than legalname to preserve backwards compatibility.
+    preferredname: `${prefix}/preferredname`,
```
Comment (toolness) [inline frontend/lib/account-settings/route-info.ts:8]: Our typical naming convention for URL paths is to hyphenate where we'd normally have spaces, so `preferred-name` is preferable here to `preferredname`, but it's not that big a deal.

**Review by toolness — APPROVED** [review]
This looks good!

One thing I noticed is that we don't mention anything about the preferred first name being optional:

> image

In other parts of the UX, we have an `(optional)` text after the label to indicate such, e.g. in EFNY we have:

> image

Should we have something similar next to "Preferred first name"?

Anyways, I have one other comment I think, but nothing is a blocker.  Feel free to merge--and great job on this epic PR!

**Comment by toolness** [comment]
Oh, one thing we'll need to figure out in the future, which I don't think is reflected in this PR, is how to express a preferred first name to TextIt.  Right now I believe we give them the user's full legal name, but I think that's not actually ideal, because most of the time we're only using their first name in informal contexts, e.g. for our rent history follow-up flow:

> image

The way things are set up right now, I _think_ we'd be addressing the user by their _legal_ first name here, rather than their preferred one, even though the informal social context is such that their preferred one is more appropriate.  I'm not even sure there are any situations we'd want to use their legal first name from TextIt, so it might be worth just giving TextIt the user's full _preferred_ name rather than their full legal one...
